kernel/src/main/scala/cats/kernel/instances/FunctionInstances.scala

```diff
@@ -127,11 +144,26 @@ trait Function1Group[A, B] extends Function1Monoid[A, B] with Group[A => B] {
     (a: A) => B.inverse(x(a))
 }
 
+final private[instances] case class CombineFunction0[A](left: () => A, right: () => A, semiA: Semigroup[A])
+    extends (() => A) {
+  private def call(fn: () => A): TailRec[A] =
+    fn match {
+      case CombineFunction0(l, r, sg) =>
+        for {
+          la <- tailcall(call(l))
+          ra <- tailcall(call(r))
+        } yield sg.combine(la, ra)
+      case _ => done(fn())
+    }
+
+  final override def apply(): A = call(this).result
+}
+
 trait Function0Semigroup[A] extends Semigroup[() => A] {
   implicit def A: Semigroup[A]
 
   override def combine(x: () => A, y: () => A): () => A =
-    () => A.combine(x(), y())
+    CombineFunction0(x, y, A)
 }
```

Comment:
can we also add `combineAllOption` here?

Reply:
Addressed